src/java.base/unix/native/jspawnhelper/jspawnhelper.c

```diff
@@ -148,8 +165,15 @@ int main(int argc, char *argv[]) {
 #ifdef DEBUG
     jtregSimulateCrash(0, 4);
 #endif
-    r = sscanf (argv[1], "%d:%d:%d", &fdinr, &fdinw, &fdout);
-    if (r == 3 && fcntl(fdinr, F_GETFD) != -1 && fcntl(fdinw, F_GETFD) != -1) {
+    r = sscanf (argv[1], "%d:%d:%d:%d:%d:%d:%d", &fdinr, &fdinw, &fdout, &jdk_feature, &jdk_interim, &jdk_update, &jdk_patch);
+    if (r == 7 && fcntl(fdinr, F_GETFD) != -1 && fcntl(fdinw, F_GETFD) != -1) {
+        // Check that JDK version and jspawnhelper version are the same
+        if (jdk_feature != VERSION_FEATURE || jdk_interim != VERSION_INTERIM || jdk_update != VERSION_UPDATE || jdk_patch != VERSION_PATCH) {
+            fprintf(stderr, "Expected jspawnhelper for Java %d.%d.%d.%d, ", jdk_feature, jdk_interim, jdk_update, jdk_patch);
+            fprintf(stderr, "but jspawnhelper for Java %d.%d.%d.%d was found.\n", VERSION_FEATURE, VERSION_INTERIM, VERSION_UPDATE, VERSION_PATCH);
```

Comment:
Minor: It is a bit odd to see that jspawnhelper found its own version odd. I'd say:

```
fprintf(stderr, "jspawnhelper version check failed. jspawnhelper version: %d.%d.%d.%d, JDK version: %d.%d.%d.%d\n",
        VERSION_FEATURE, VERSION_INTERIM, VERSION_UPDATE, VERSION_PATCH,
        jdk_feature, jdk_interim, jdk_update, jdk_patch);
```